crates/net/downloaders/src/optimistic/mod.rs

```diff
@@ -0,0 +1,246 @@
+use futures::{Future, FutureExt, StreamExt};
+use reth_interfaces::{
+    consensus::Consensus,
+    p2p::{bodies::client::BodiesClient, headers::client::HeadersClient},
+};
+use reth_primitives::H256;
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio_stream::wrappers::UnboundedReceiverStream;
+use tracing::*;
+
+/// Downloader config.
+pub mod config;
+use config::{BatchDownloadConfig, DownloaderConfig};
+
+/// Downloader error.
+pub mod error;
+pub use error::{DownloaderError, DownloaderResult};
+
+/// Downloader incoming message.
+pub mod message;
+use message::DownloaderMessage;
+
+mod download;
+use download::{bodies::BodiesDownload, headers::HeadersDownload, Download, DownloadOrigin};
+
+mod buffer;
+use buffer::DownloadBuffer;
+
+mod state;
+use state::DownloaderState;
+
+mod utils;
+pub use utils::{BoundedBinaryHeap, BoundedVec};
+
+/// Optimistic downloader implementation.
+#[derive(Debug)]
+pub struct Downloader<N, C> {
+    /// The network client
+    client: Arc<N>,
+    /// The consensus client
+    consensus: Arc<C>,
+    /// Chain tip
+    tip: H256, // TODO: move to state
+    /// Header batch size.
+    config: BatchDownloadConfig,
+    /// The buffer of download results.
+    buffer: DownloadBuffer,
+    /// The request receiver.
+    message_rx: UnboundedReceiverStream<DownloaderMessage>,
+    /// Pending downloads.
+    downloads: BoundedVec<Download<N, C>>,
+    /// The download queue.
+    queue: BoundedBinaryHeap<Download<N, C>>,
+}
+
+impl<N, C> Downloader<N, C>
+where
+    N: HeadersClient + 'static,
+    C: Consensus,
+{
+    /// Create new instance of downloader.
+    pub fn new(
+        client: Arc<N>,
+        consensus: Arc<C>,
+        config: DownloaderConfig,
+        request_rx: UnboundedReceiverStream<DownloaderMessage>,
```

Comment:
why does this take a receiver?

how will this be used inside the stage?

Reply:
similar to a `NetworkHandle`, i'll create a `DownloaderHandle` which would be able to forward requests to the downloader